### PR TITLE
Extends STM32 OSPI NOR-Flash support

### PR DESCRIPTION
**Pull request by GeorgeCGV**
- Extends OSPI include defines with dual and quad modes
- Extends STM32 OSPI driver:
  - Automatically detects read opcode from SFDP: BFP.
    Supports the following modes:
     - `1-1-2`
     - `1-2-2`
     - ~`2-2-2`~
     - `1-1-4`
     - `1-4-4`
     - ~`4-4-4`~
  - Allows setting desired write/program opcode with `writeoc` DTS property (*not for OPI mode*).
    If absent, the driver uses the PP opcode based on the configured mode (*see the bindings*). 
    But the following PP opcodes are supported:
    - `0x02`
    - `0xA2`
    - `0x32`
    - `0x38`
  - Can use 4-Byte address opcodes if NOR-Flash (*if required*).
    The option can be enabled in DTS, available as `four-byte-opcodes`.
    Note: it automatically converts PP & Read opcodes. However, not the erase opcode(s).
  - Supports `quad-enable-requirements` DTS property defined in `jedec,jesd216`.
    Parses SFDP: BFP DW15 and bypasses the DTS when parsing is successful.
    The driver will automatically detect the required QER and enable it when configured in quad mode.

Tested against:

<details>
  <summary>W25Q128JVSIQ</summary>

**DTS**

```
ospi_flash: ospi-nor-flash@0 {
    status = "okay";
    compatible = "st,stm32-ospi-nor";
    label = "W25Q128JVSIQ";
    reg = <0>;
    ospi-max-frequency = <DT_FREQ_M(90)>;
    size = <DT_SIZE_M(16*8)>;
    spi-bus-width = <OSPI_QUAD_MODE>;
    data-rate = <OSPI_STR_TRANSFER>;
    writeoc = "PP_1_1_4";
};
```

**Flash Test Output**

```
uart:~$ flash test W25Q128JVSIQ 0 0x1000 3
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK.
```

</details>

<details>
  <summary>MX25L12835F</summary>

**DTS**

```
ospi_flash: ospi-nor-flash@0 {
    status = "okay";
    compatible = "st,stm32-ospi-nor";
    label = "MX25L12835F";
    reg = <0>;
    ospi-max-frequency = <DT_FREQ_M(90)>;
    size = <DT_SIZE_M(16*8)>;
    spi-bus-width = <OSPI_QUAD_MODE>;
    data-rate = <OSPI_STR_TRANSFER>;
};
```

**Flash Test Output**

```
uart:~$ flash test MX25L12835F 0 0x1000 3
Erase OK.
Write OK.
Erase OK.
Write OK.
Erase OK.
Write OK. 
```

</details>

Using custom H730 board. Therefore, the PR required #45732. However, it is updated to be independent.

The OSPI driver supports more modes than the QSPI driver. Eventually, the content of both drivers shall be refactored to extract common part(s). I suggest doing that after ongoing PRs are merged:

* #45732
* #45487
* #45151

### CODE REVIEW

**Comment by GeorgeCGV** [comment]
Removed dependency on #45732.
Corrected to limit read mode based on `data_mode`. For example, that allows the OSPI to be limited to dual read mode only.

**Comment by manoj153** [comment]
> @manoj153, considering that quad mode works fine for `W25Q128JVSIQ` and `MX25L12835F` it is interesting to know if that is the driver issue or the HW regarding the `W25Q32JVSFIQ`.
> 
> Are you able to operate the `W25Q32JVSFIQ` without the driver using the STM32 OSPI example or generated STM32Cube project?

I concur, 

Nope, I have not tested with another codebase, I will be going to test with a standalone HAL code and see if I will be having issues.

Thanks for the suggestion.

**Review by mbolivar-nordic — COMMENTED** [review]
DT bindings changes look good to me. I leave it to the STM32 maintainers to provide the approval, however.

File: dts/bindings/flash_controller/st,stm32-ospi-nor.yaml
Comment (nvlsianpu) [inline dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:74]: if you look into `dts/bindings/mtd/nordic,qspi-nor.yaml` file you can see different notation for the same kind of property.

Comment (nvlsianpu) [reply dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:74]: I agree on that you can use different solution in this PR. I just want to highlight other notation.

I would say that option 1. is outstanding IMO.
And yes this PR shouldn't be extended for reworking the others, so that topic for other patch as you mentioned.

**Comment by GeorgeCGV** [comment]
Converted to draft as I discovered some cases when flash ICs SFDP mode support contradicts its reference manual.

For example: `W25Q128JVSIQ` SFDP `DW5` bit `4` is `1`, which indicates that it supports 4-4-4 mode. Unfortunately, in the reference manual, the support of the `4-4-4` mode is not mentioned.

One approach to fix that is to allow read mode to be set in the DTS. Then SFDP read mode can be ignored or used to double-check if the mode is supported and to get its instruction with dummy cycles.

File: dts/bindings/flash_controller/st,stm32-ospi-nor.yaml
Comment (GeorgeCGV) [reply dts/bindings/flash_controller/st,stm32-ospi-nor.yaml:74]: Considering the recent discovery... It seems that the `mode` doesn't really define the instruction and flash can use any instruction it likes for any mode. It is just a common practice to use certain opcodes for certain modes.

Therefore, instead of the read and write mode fields it shall be read and write opcodes with an extra dummy cycles field for the read.

**Comment by GeorgeCGV** [comment]
Rebased, removed support of `2-2-2` and `4-4-4` read modes. At the moment, that is the most pragmatic way. Considering that they are not the most used modes.

**Review by nvlsianpu — APPROVED** [review]
+1 to spi_nor.h and dt bindings.

**Review by erwango — APPROVED** [review]

**Pull request merged**
